lib/questionlib.php

```diff
@@ -2594,3 +2076,726 @@ function get_next_version(int $questionbankentryid): ?int {
+     *
+     * @param string $cap capability
+     * @return boolean
+     */
+    public function have_cap($cap) {
+        return (count($this->having_cap($cap)));
+    }
+
+    /**
+     * Has at least one parent context got one of the caps $caps?
+     *
+     * @param array $caps capability
+     * @return boolean
+     */
+    public function have_one_cap($caps) {
+        foreach ($caps as $cap) {
+            if ($this->have_cap($cap)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Has at least one parent context got one of the caps for actions on $tabname
+     *
+     * @param string $tabname edit tab name
+     * @return boolean
+     */
+    public function have_one_edit_tab_cap($tabname) {
+        return $this->have_one_cap(self::$caps[$tabname]);
+    }
+
+    /**
+     * Throw error if at least one parent context hasn't got the cap $cap
+     *
+     * @param string $cap capability
+     */
+    public function require_cap($cap) {
+        if (!$this->have_cap($cap)) {
+            throw new \moodle_exception('nopermissions', '', '', $cap);
+        }
+    }
+
+    /**
+     * Throw error if at least one parent context hasn't got one of the caps $caps
+     *
+     * @param array $caps capabilities
+     */
+    public function require_one_cap($caps) {
+        if (!$this->have_one_cap($caps)) {
+            $capsstring = join(', ', $caps);
+            throw new \moodle_exception('nopermissions', '', '', $capsstring);
+        }
+    }
+
+    /**
+     * Throw error if at least one parent context hasn't got one of the caps $caps
+     *
+     * @param string $tabname edit tab name
+     */
+    public function require_one_edit_tab_cap($tabname) {
+        if (!$this->have_one_edit_tab_cap($tabname)) {
+            throw new \moodle_exception('nopermissions', '', '', 'access question edit tab '.$tabname);
+        }
+    }
+
+    /**
+     * Remove stale questions from a category.
+     *
+     * While questions should not be left behind when they are not used any more,
+     * it does happen, maybe via restore, or old logic, or uncovered scenarios. When
+     * this happens, the users are unable to delete the question category unless
+     * they move those stale questions to another one category, but to them the
+     * category is empty as it does not contain anything. The purpose of this function
+     * is to detect the questions that may have gone stale and remove them.
+     *
+     * You will typically use this prior to checking if the category contains questions.
+     *
+     * The stale questions (unused and hidden to the user) handled are:
+     * - hidden questions
+     * - random questions
+     *
+     * @param int $categoryid The category ID.
+     * @deprecated since Moodle 4.0 MDL-71585
+     * @see qbank_managecategories\helper
+     */
+    function question_remove_stale_questions_from_category($categoryid) {
```

Comment:
major problem from point, these functions are moved inside question_edit_contexts class, need to be fixed